orderbook/src/account_balances.rs

```diff
@@ -45,73 +56,74 @@ pub struct Web3BalanceFetcher {
+    source: SellTokenSource,
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone, Debug, Default)]
 struct SubscriptionValue {
     balance: Option<U256>,
     allowance: Option<U256>,
 }
 
 impl Web3BalanceFetcher {
-    pub fn new(web3: Web3, vault_relayer: H160, settlement_contract: H160) -> Self {
+    pub fn new(
+        web3: Web3,
+        vault: Option<BalancerV2Vault>,
+        vault_relayer: H160,
+        settlement_contract: H160,
+    ) -> Self {
         Self {
             web3,
+            vault,
             vault_relayer,
             settlement_contract,
             balances: Default::default(),
         }
     }
 
-    async fn _register_many(
-        &self,
-        subscriptions: impl Iterator<Item = SubscriptionKey>,
-    ) -> Result<()> {
+    async fn _register_many(&self, subscriptions: Vec<SubscriptionKey>) -> Result<()> {
         let mut batch = CallBatch::new(self.web3.transport());
 
         // Make sure subscriptions are registered for next update even if batch call fails
-        let subscriptions: Vec<SubscriptionKey> = {
+        {
             let mut guard = self.balances.lock().expect("thread holding mutex panicked");
-            subscriptions
-                .map(|subscription| {
-                    let _ = guard.entry(subscription).or_default();
-                    subscription
-                })
-                .collect()
-        };
+            for subscription in &subscriptions {
+                guard.insert(*subscription, Default::default());
+            }
+        }
 
         let calls = subscriptions
             .into_iter()
             .map(|subscription| {
-                let instance = ERC20::at(&self.web3, subscription.token);
-                join3(
-                    instance
-                        .balance_of(subscription.owner)
-                        .batch_call(&mut batch),
-                    instance
-                        .allowance(subscription.owner, self.vault_relayer)
-                        .batch_call(&mut batch),
-                    std::future::ready(subscription),
-                )
+                let token = ERC20::at(&self.web3, subscription.token);
+                let value = match (subscription.source, &self.vault) {
+                    (SellTokenSource::Erc20, _) => erc20_balance_query(
+                        &mut batch,
+                        token,
+                        subscription.owner,
+                        self.vault_relayer,
+                    ),
+                    (SellTokenSource::External, Some(vault)) => vault_external_balance_query(
+                        &mut batch,
+                        vault.clone(),
+                        token,
+                        subscription.owner,
+                        self.vault_relayer,
+                    ),
+                    _ => async { SubscriptionValue::default() }.boxed(),
+                };
+                future::join(future::ready(subscription), value)
```

Comment:
Alternatively we could have one big async block inside of which we do the match and at the end return the tuple (subscription, value). That way no boxing is needed because the futures all have the same type.

Reply:
Boxing felt easier as depending on the balance type we need to add different calls to the batch.